h5totxt.1
.\" Copyright (c) 1999-2009 Massachusetts Institute of Technology
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be
.\" included in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\" EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
.\" CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
.\" TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
.\" SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.TH H5TOTXT 1 "March 9, 2002" "h5utils" "h5utils"
.SH NAME
h5totxt \- generate comma-delimited text from 2d slices of HDF5 files
.SH SYNOPSIS
.B h5totxt
[\fIOPTION\fR]... [\fIHDF5FILE\fR]...
.SH DESCRIPTION
.PP
." Add any additional description here
h5totxt is a utility to generate comma-delimited text (and similar
formats) from one-, two-, or more-dimensional slices of numeric
datasets in HDF5 files.  This way, the data can easily be imported
into spreadsheets and similar programs for analysis and visualization.

HDF5 is a free, portable binary format and supporting library developed
by the National Center for Supercomputing Applications at the University
of Illinois in Urbana-Champaign.  A single
.I h5
file can contain multiple data sets; by default,
.I h5totxt
takes the first dataset, but this can be changed via the
.B -d
option, or by using the syntax \fIHDF5FILE:DATASET\fR.

By default, the entire dataset is dumped to the output.  in row-major
order.  For 3d datasets, this corresponds to a sequence of yz slices,
in order of increasing x, separated by blank lines.  If
.B -T
is specified, outputs in the transposed (column-major) order instead

Often, however, you want only a one- or two-dimensional slice of
multi-dimensional data.  To do this, you specify coordinates in one or
more slice dimensions, via the
.B -xyzt
options.

The most basic usage is something like \'h5totxt foo.h5\', which will
output comma-delimited text to stdout from the data in foo.h5.
.SH OPTIONS
.TP
.B -h
Display help on the command-line options and usage.
.TP
.B -V
Print the version number and copyright info for h5totxt.
.TP
.B -v
Verbose output.
.TP
\fB\-o\fR \fIfile\fR
Send text output to
.I file
rather than to stdout (the default).
.TP
\fB\-s\fR \fIsep\fR
Use the string
.I sep
to separate columns of the output rather than a comma (the default).
.TP
\fB\-x\fR \fIix\fR, \fB\-y\fR \fIiy\fR, \fB\-z\fR \fIiz\fR, \fB\-t\fR \fIit\fR
This tells
.I h5totxt
to use a particular slice of a multi-dimensional dataset.  e.g.
.B -x
causes a yz plane (of a 3d dataset) to be used, at an x index of
.I ix
(where the indices run from zero to one less than the maximum index in
that direction).  Here, x/y/z correspond to the first/second/third
dimensions of the HDF5 dataset. The \fB\-t\fR option specifies a slice
in the last dimension, whichever that might be.  See also the
.B -0
option to shift the origin of the x/y/z slice coordinates to the
dataset center.
.TP
.B -0
Shift the origin of the x/y/z slice coordinates to the dataset center,
so that e.g. -0 -x 0 (or more compactly -0x0) returns the central x
plane of the dataset instead of the edge x plane.  (\fB\-t\fR
coordinates are not affected.)
.TP
.B -T
Transpose the data (interchange the dimension ordering).  By default, no
transposition is done.
.TP
\fB\-.\fR \fInumdigits\fR
Output
.I numdigits
digits after the decimal point (defaults to 16).
.TP
\fB\-d\fR \fIname\fR
Use dataset
.I name
from the input files; otherwise, the first dataset from each file is used.
Alternatively, use the syntax \fIHDF5FILE:DATASET\fR, which allows you
to specify a different dataset for each file.
You can use the
.I h5ls
command (included with hdf5) to find the names of datasets within a file.
.SH BUGS
Send bug reports to S. G. Johnson, stevenj@alum.mit.edu.
.SH AUTHORS
Written by Steven G. Johnson.  Copyright (c) 2005 by the Massachusetts
Institute of Technology.


NEWS
h5utils 1.12 (6/12/09)

  * The vertical axis in h5topng is now reversed to correspond to
    what most people seem to expect: increasing coordinates correspond
    to "up" and "right" in the image, rather than "down" and "right"
    in the image as in previous versions.

  * Fixed failure in h5tovtk -2; thanks to Karen Lee for the bug report.

  * Fixed installation of h5read.oct for Octave 3.x.

h5utils 1.11.1 (4/28/08)

  * Fixed failure to find colormap files in h5topng 1.11 (due
    to changes in autoconf 2.60); thanks to bug report from Jiangjun Zheng.

h5utils 1.11 (4/24/08)

  * h5tovtk no longer reverses the dimensions; thanks to Andreas Wilde
    for the suggestion.

  * Fix compilation failure with HDF5 1.8.

h5utils 1.10.1 (9/20/06)

  * Fixed build problem on Cygwin due to missing ".exe" extension. 
    Thanks to Ken Hill for the bug report.

h5utils 1.10 (9/2/05)

  * Added h4fromh5 utility.  (NCSA seems to be no longer shipping
    the HDF5/HDF4 conversion tools with the latest HDF5 release.)

  * Added dkbluered color map, which is similar to bluered but
    uses a somewhat wider range of colors.

h5utils 1.9.1 (8/5/04)

  * Fix h5topng compilation failure with some non-C99 compilers;
    thanks to Maarten van Reeuwijk for the bug report.

h5utils 1.9 (7/12/04)

  * Added new h5math utility, which creates and combines HDF5 datasets
    using a user-specified mathematical expression.  (Requires GNU
    libmatheval to be installed.)

  * h5topng: Matlab-like start:end or start:step:end notation for
    slice indices, to allow a whole sequence of slices to be output as
    a sequence of PNG images.

  * h5topng: if contour/overlay dataset does not have same dimensions
    as output data, it is periodically "tiled" over the output.

h5utils 1.8 (5/22/04)

  * New -A and -a options for h5topng to allow translucent overlays
    from one file onto another (an alternative or complement to the
    -C contour-overlay option).

  * h5topng uses 24-bit direct color by default (use -8 option for
    old 8-bit behavior).

  * h5topng uses columns/rows for x/y by default (use -T to swap),
    the opposite of the old behavior.

  * There is no default -z 0 slice dimension in h5topng/h5totxt any more.
    You must specify a slice for 3+ dimensional data in h5topng.
    h5totxt dumps the whole data file by default unless one or more
    slices are specified.

  * Support specifying multiple slice dimensions for 4+ dimensional
    datasets, with new "-t" option to indicate final dimension.

  * Slices are also supported now in h5tovtk and h5tov5d.

  * New -. option in h5totxt to specify number of significant digits; 
    output 16 digits by default instead of 6, previously.

h5utils 1.7.1 (3/16/02)

  * Fixed array overrun in h5topng that caused a floating-point exception
    on Alphas; thanks to Marin Soljacic for the bug report.

h5utils 1.7 (3/9/02)

  * h5topng now supports multiple, user-definable color tables, a number
    of which are provided.  INCOMPATIBLE CHANGE: the -c option now has
    the syntax: -c <colortable>.  The old behavior corresponds to the
    included "bluered" colortable, invoked via: -c bluered.

  * New -R option for h5topng to use a consistent color scale for
    all specified files.

  * New -0 option for h5topng and h5totxt that shifts the origin of the
    slice coordinates to the dataset center.

  * Added h5tovtk program to output VTK (Visualization ToolKit) data files.

  * Support -T (transpose dimensions) option in h5tov5d.

  * Fixed bug in h5topng that caused extra rows/columns of garbage pixels
    to be written at the edges of images when scaling was used.

  * When compiling the h5read Octave plugin, respect the CPPFLAGS and LDFLAGS 
    environment variables.  Thanks to Max Colice for the bug report.

  * Fixed problem when --without-h5tov5d and --without-h5fromh4 are used. 
    Thanks to Nikola Ivanov Nikolov for the fix.

h5utils 1.6 (1/17/01)

  * Don't build h5fromh4 if the superior h4toh5 tool (from HDF5 1.4)
    is present.  Also added --{with,without}-h5fromh4 option to
    configure to force whether h5fromh4 is built.

h5utils 1.5.1 (12/9/00)

  * Support manually disabling Octave plugin support
    (configure --without-octave) in case of C++ problems.

  * Support Vis5d+ and Debian HDF header file locations.

h5utils 1.5 (7/9/00)

  * Added h5fromh4 program to convert HDF4 datasets to HDF5 format.

  * Added -S <s> option to h5topng as a shortcut for -X <s> -Y <s>.

h5utils 1.4 (5/28/00)

  * Added h5fromtxt program to convert text input to an HDF5 dataset.

h5utils 1.3.4 (1/31/00)

  * Improved -C contour plotting in h5topng.

  * Fix in h5topng man page (thanks to Christoph Becher).

h5utils 1.3.3 (1/30/00)

  * Bug fix in h5topng (would sometimes output solid black images).
    Thanks to Karl Koch for the bug report.

h5utils 1.3.2 (1/28/00)

  * Added h5topng -Z option to center color scale on zero.

  * Now support h5topng -C filename:dataset.

h5utils 1.3.1 (1/27/00)

  * Bug fixes in dataset name-handling, especially when using h5tov5d to
    join multiple datasets into one output file.

h5utils 1.3 (1/21/00)

  * You can now specify individual datasets within a file by using
    <filename>:<dataset> instead of just <filename> with h5topng,
    h5totxt, and h5tov5d.

h5utils 1.2.3: (1/20/00)

  * Fixed minor bug in 'h5totxt -h'.

h5utils 1.2.2: (1/12/00)

  * Makefile now includes CPPFLAGS in the compiler flags, making it
    easier to use header files in non-standard locations.  (CPPFLAGS
    is the proper place to put -I flags for the configure script.)

h5utils 1.2.1:  (1/1/00)

  * Modified -o option of h5tov5d to join datasets into a single Vis5d file.

h5utils 1.2:  (12/31/99)

  * Added h5tov5d program for converting to Vis5d format.

  * Improved installation; h5read.oct now goes into the sitewide Octave
    plugins directory, and things work correctly when only some of the
    utilities are compiled.


h5totxt.c
/* Copyright (c) 1999-2009 Massachusetts Institute of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <unistd.h>

#include "config.h"
#include "arrayh5.h"
#include "copyright.h"
#include "h5utils.h"

#define CHECK(cond, msg) { if (!(cond)) { fprintf(stderr, "h5totxt error: %s\n", msg); exit(EXIT_FAILURE); } }

void usage(FILE *f)
{
     fprintf(f, "Usage: h5totxt [options] [<filenames>]\n"
	     "Options:\n"
	     "         -h : this help message\n"
             "         -V : print version number and copyright\n"
	     "         -v : verbose output\n"
	     "   -s <sep> : use <sep> to separate columns [ default: \",\" ]\n"
	     "  -o <file> : output to <file> (first input file only)\n"
	     "    -x <ix> : take x=<ix> slice of data\n"
	     "    -y <iy> : take y=<iy> slice of data\n"
	     "    -z <iz> : take z=<iz> slice of data\n"
	     "    -t <it> : take t=<it> slice of data's last dimension\n"
	     "         -0 : use dataset center as origin for -x/-y/-z\n"
	     "         -T : transpose the data [default: no]\n"
	     "     -. <n> : output <n> decimal places [ default: 16 ]\n"
	     "  -d <name> : use dataset <name> in the input files (default: first dataset)\n"
	     "              -- you can also specify a dataset via <filename>:<name>\n"
	  );
}

int main(int argc, char **argv)
{
     arrayh5 a;
     char *txt_fname = NULL, *data_name = NULL;
     extern char *optarg;
     extern int optind;
     int c;
     int slicedim[4] = {NO_SLICE_DIM,NO_SLICE_DIM,NO_SLICE_DIM,NO_SLICE_DIM};
     int islice[4], center_slice[4] = {0,0,0,0};
     int err;
     int nx, ny, nz;
     int dec = 16;
     int verbose = 0;
     int transpose = 0;
     char *sep;
     int ifile;

     sep = my_strdup(",");

     while ((c = getopt(argc, argv, "ho:x:y:z:t:0ad:vTs:.:V")) != -1)
	  switch (c) {
	      case 'h':
		   usage(stdout);
		   return EXIT_SUCCESS;
	      case 'V':
		   printf("h5totxt " PACKAGE_VERSION " by Steven G. Johnson\n" 
			  COPYRIGHT);
		   return EXIT_SUCCESS;
	      case 'v':
		   verbose = 1;
		   break;
              case '0':
                   center_slice[0] = center_slice[1] = center_slice[2] = 1;
                   break;
	      case 'T':
		   transpose = 1;
		   break;
	      case 'o':
		   free(txt_fname);
                   txt_fname = my_strdup(optarg);
                   break;
	      case 's':
		   free(sep);
		   sep = my_strdup(optarg);
		   break;
	      case 'd':
		   free(data_name);
		   data_name = my_strdup(optarg);
		   break;		   
	      case '.':
		   dec = atoi(optarg);
		   break;
	      case 'x':
		   islice[0] = atoi(optarg);
		   slicedim[0] = 0;
		   break;
	      case 'y':
		   islice[1] = atoi(optarg);
		   slicedim[1] = 1;
		   break;
	      case 'z':
		   islice[2] = atoi(optarg);
		   slicedim[2] = 2;
		   break;
	      case 't':
		   islice[3] = atoi(optarg);
		   slicedim[3] = LAST_SLICE_DIM;
		   break;
	      case 'a':
		   slicedim[0] = slicedim[1] = slicedim[2] = slicedim[3]
			= NO_SLICE_DIM;
		   break;
	      default:
		   fprintf(stderr, "Invalid argument -%c\n", c);
		   usage(stderr);
		   return EXIT_FAILURE;
	  }
     if (optind == argc) {  /* no parameters left */
	  usage(stderr);
	  return EXIT_FAILURE;
     }

     for (ifile = optind; ifile < argc; ++ifile) {
	  char *dname, *h5_fname;
	  h5_fname = split_fname(argv[ifile], &dname);
	  if (!dname[0])
	       dname = data_name;

	  if (verbose) {
	       int i;
	       printf("reading from \"%s\"", h5_fname);
	       for (i = 0; i < 4; ++i)
		    if (slicedim[i] != NO_SLICE_DIM)
			 printf(", slice at %d in %c dimension", islice[i], 
				slicedim[i] == LAST_SLICE_DIM ? 't' 
				: slicedim[i] + 'x');
	       printf(".\n");
	  }
	  
	  err = arrayh5_read(&a, h5_fname, dname, NULL,
			     4, slicedim, islice, center_slice);
	  CHECK(!err, arrayh5_read_strerror[err]);

	  if (transpose)
	       arrayh5_transpose(&a);

	  {
	       double a_min, a_max;
	       arrayh5_getrange(a, &a_min, &a_max);
	       if (verbose)
		    printf("data ranges from %.*g to %.*g.\n",
			   dec, a_min, dec, a_max);
	  }
	  
	  nx = a.rank < 1 ? 1 : a.dims[0];
	  ny = a.rank < 2 ? 1 : a.dims[1];
	  nz = a.rank < 3 ? 1 : a.dims[2];
	  
	  if (verbose && a.rank <= 3)
	       printf("writing %s from %dx%dx%d input data.\n",
		      txt_fname ? txt_fname : "to stdout", nx, ny, nz);

	  {
	       FILE *f;
	       int i, j, k;

	       if (txt_fname) {
		    f = fopen(txt_fname, "w");
		    CHECK(f, "error creating file");
	       }
	       else
		    f = stdout;
	       
	       if (a.rank < 3)
		    for (i = 0; i < nx; ++i) {
			 if (ny > 0)
			      fprintf(f, "%.*g", dec, a.data[i*ny + 0]);
			 for (j = 1; j < ny; ++j)
			      fprintf(f, "%s%.*g", sep, dec, a.data[i*ny + j]);
			 fprintf(f, "\n");
		    }
	       else if (a.rank == 3)
		    for (i = 0; i < nx; ++i) {
			 if (i > 0)
			      fprintf(f, "\n");
			 for (j = 0; j < ny; ++j) {
			      int ij = nz * (ny * i + j);
			      if (nz > 0)
				   fprintf(f, "%.*g", dec, a.data[ij + 0]);
			      for (k = 1; k < nz; ++k)
				   fprintf(f, "%s%.*g",
					   sep, dec, a.data[ij + k]);
			      fprintf(f, "\n");
			 }
		    }
	       else {
		    if (a.N > 0)
			 fprintf(f, "%.*g", dec, a.data[0]);
		    for (i = 0; i < a.N; ++i) {
			 fprintf(f, "%s%.*g", sep, dec, a.data[i]);
		    }
		    fprintf(f, "\n");
	       }

	       if (txt_fname)
		    fclose(f);
	  }

	  arrayh5_destroy(a);
	  if (txt_fname)
	       free(txt_fname);
	  txt_fname = NULL;
	  free(h5_fname);
     }
     free(sep);
     free(data_name);

     return EXIT_SUCCESS;
}
